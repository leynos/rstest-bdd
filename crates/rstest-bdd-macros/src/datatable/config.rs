use proc_macro2::Ident;
use syn::{ExprPath, Type};

use super::rename::RenameRule;

pub(crate) struct StructConfig {
    pub(crate) rename_rule: Option<RenameRule>,
}

pub(crate) enum DefaultValue {
    Trait,
    Function(ExprPath),
}

#[derive(Clone)]
pub(crate) enum Accessor {
    Column { name: String },
    Index { position: usize },
}

pub(crate) struct FieldConfig {
    pub(crate) accessor: Accessor,
    pub(crate) optional: bool,
    pub(crate) default: Option<DefaultValue>,
    pub(crate) parse_with: Option<ExprPath>,
    pub(crate) truthy: bool,
    pub(crate) trim: bool,
}

impl FieldConfig {
    pub(crate) fn new(accessor: Accessor) -> Self {
        Self {
            accessor,
            optional: false,
            default: None,
            parse_with: None,
            truthy: false,
            trim: false,
        }
    }
}

pub(crate) struct FieldSpec {
    pub(crate) ident: Option<Ident>,
    pub(crate) ty: Type,
    pub(crate) inner_ty: Type,
    pub(crate) config: FieldConfig,
}
