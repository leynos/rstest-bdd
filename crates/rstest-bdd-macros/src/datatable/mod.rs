//! Derive entry points for typed data table helpers.
//!
//! This module wires the outer derive macro surfaces to the internal code
//! generators that live in the sibling modules.

mod config;
mod parser;
mod rename;
mod row;
mod table;
mod validation;

use proc_macro::TokenStream;

pub(crate) fn derive_data_table_row(input: TokenStream) -> TokenStream {
    row::expand(input)
}

pub(crate) fn derive_data_table(input: TokenStream) -> TokenStream {
    table::expand(input)
}
