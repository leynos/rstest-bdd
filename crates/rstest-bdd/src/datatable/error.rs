//! Error types surfaced by the datatable runtime.

use std::error::Error as StdError;

use thiserror::Error;

/// Errors that can arise when converting a Gherkin data table into typed rows.
///
/// `DataTableError` values surface through the generated step wrappers,
/// ensuring that failures are reported with helpful row and column context.
#[derive(Debug, Error)]
#[non_exhaustive]
pub enum DataTableError {
    /// Raised when a typed parser expects a header row but the table is empty.
    #[error("data table requires a header row")]
    MissingHeader,
    /// Raised when the header row repeats a column name.
    #[error("data table header contains duplicate column '{column}'")]
    DuplicateHeader { column: String },
    /// Raised when a row contains more or fewer cells than expected.
    #[error("data table row {row_number} has {actual} cells but expected {expected}")]
    UnevenRow {
        /// 1-based index of the row that failed, including any header.
        row_number: usize,
        /// Number of cells required for each row.
        expected: usize,
        /// Number of cells present in the offending row.
        actual: usize,
    },
    /// Raised when a column name lookup fails.
    #[error("data table row {row_number} is missing column '{column}'")]
    MissingColumn {
        /// 1-based index of the row that failed, including any header.
        row_number: usize,
        /// Name of the column that was requested.
        column: String,
    },
    /// Raised when positional access references an out-of-range column index.
    #[error("data table row {row_number} is missing cell {column_index}")]
    MissingCell {
        /// 1-based index of the row that failed, including any header.
        row_number: usize,
        /// 1-based index of the missing column.
        column_index: usize,
    },
    /// Raised when parsing an entire row fails.
    #[error("row {row_number}: {source}")]
    RowParse {
        /// 1-based index of the row that failed, including any header.
        row_number: usize,
        /// Root cause reported by the row parser.
        #[source]
        source: Box<dyn StdError + Send + Sync>,
    },
    /// Raised when parsing an individual cell fails.
    #[error("row {row_number}, column {column_index}{column_label}: {source}")]
    CellParse {
        /// 1-based index of the row that failed, including any header.
        row_number: usize,
        /// 1-based index of the column that failed to parse.
        column_index: usize,
        /// Display-formatted column label used when a header was present.
        column_label: String,
        /// Root cause reported by the cell parser.
        #[source]
        source: Box<dyn StdError + Send + Sync>,
    },
}

impl DataTableError {
    pub(crate) fn cell_parse<E>(
        row_number: usize,
        column_index: usize,
        column_label: Option<String>,
        err: E,
    ) -> Self
    where
        E: StdError + Send + Sync + 'static,
    {
        let column_label = column_label
            .map(|label| format!(" ({label})"))
            .unwrap_or_default();
        Self::CellParse {
            row_number,
            column_index: column_index + 1,
            column_label,
            source: Box::new(err),
        }
    }
}
