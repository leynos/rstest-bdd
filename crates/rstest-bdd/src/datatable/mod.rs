//! Runtime helpers for working with typed Gherkin data tables.
mod error;
mod parsers;
mod rows;
mod spec;

pub use error::DataTableError;
pub use parsers::{trimmed, truthy_bool, TrimmedParseError, TruthyBoolError};
pub use rows::{DataTableRow, Rows};
pub use spec::{HeaderSpec, RowSpec};

#[cfg(test)]
mod tests;
