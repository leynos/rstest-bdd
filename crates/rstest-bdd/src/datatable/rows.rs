//! Typed row abstractions used when parsing datatable content.

use std::convert::TryFrom;
use std::ops::Deref;

use super::{DataTableError, HeaderSpec, RowSpec};

/// Trait implemented by types that can be constructed from a [`RowSpec`].
pub trait DataTableRow: Sized {
    /// When `true`, [`Rows<T>`](Rows) expects the first row to contain headers.
    const REQUIRES_HEADER: bool = false;

    /// Parse a row into the target type.
    ///
    /// # Errors
    ///
    /// Implementors should return [`DataTableError`] describing conversion
    /// failures.
    fn parse_row(row: RowSpec<'_>) -> Result<Self, DataTableError>;
}

/// A strongly-typed collection of parsed rows.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Rows<T>(Vec<T>);

impl<T> Rows<T> {
    pub(super) fn new(rows: Vec<T>) -> Self {
        Self(rows)
    }

    /// Returns the number of parsed rows.
    #[must_use]
    pub fn len(&self) -> usize {
        self.0.len()
    }

    /// Returns `true` when the collection is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    /// Provides shared access to the parsed rows.
    #[must_use]
    pub fn as_slice(&self) -> &[T] {
        &self.0
    }

    /// Returns an iterator over the parsed rows.
    pub fn iter(&self) -> std::slice::Iter<'_, T> {
        self.0.iter()
    }

    /// Consumes the collection, returning the underlying vector.
    #[must_use]
    pub fn into_inner(self) -> Vec<T> {
        self.0
    }
}

impl<T> IntoIterator for Rows<T> {
    type Item = T;
    type IntoIter = std::vec::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<T> Deref for Rows<T> {
    type Target = [T];

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'a, T> IntoIterator for &'a Rows<T> {
    type Item = &'a T;
    type IntoIter = std::slice::Iter<'a, T>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl<T> TryFrom<Vec<Vec<String>>> for Rows<T>
where
    T: DataTableRow,
{
    type Error = DataTableError;

    fn try_from(table: Vec<Vec<String>>) -> Result<Self, Self::Error> {
        let mut rows_iter = table.into_iter();
        let mut row_number = 1;
        let header = if T::REQUIRES_HEADER {
            let header_row = rows_iter.next().ok_or(DataTableError::MissingHeader)?;
            let spec = HeaderSpec::new(header_row)?;
            row_number += 1;
            Some(spec)
        } else {
            None
        };
        let mut parsed_rows = Vec::new();
        for (index, row) in rows_iter.enumerate() {
            if let Some(ref header) = header {
                if row.len() != header.len() {
                    return Err(DataTableError::UnevenRow {
                        row_number: row_number + index,
                        expected: header.len(),
                        actual: row.len(),
                    });
                }
            }
            let spec = RowSpec::new(header.as_ref(), row_number + index, index, row);
            let parsed = T::parse_row(spec)?;
            parsed_rows.push(parsed);
        }
        Ok(Self::new(parsed_rows))
    }
}
