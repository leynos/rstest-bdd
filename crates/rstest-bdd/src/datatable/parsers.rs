use std::error::Error as StdError;
use std::fmt;

/// Parses boolean values in a tolerant, human-friendly fashion.
///
/// `truthy_bool` recognises common affirmative and negative forms, returning a
/// [`TruthyBoolError`] when the value cannot be classified.
///
/// # Examples
/// ```
/// # use rstest_bdd::datatable::truthy_bool;
/// assert!(truthy_bool("yes").unwrap());
/// assert!(!truthy_bool("no").unwrap());
/// ```
///
/// # Errors
///
/// Returns [`TruthyBoolError`] when the input does not match a recognised form.
pub fn truthy_bool(value: &str) -> Result<bool, TruthyBoolError> {
    match value.trim().to_ascii_lowercase().as_str() {
        "yes" | "y" | "true" | "1" => Ok(true),
        "no" | "n" | "false" | "0" => Ok(false),
        other => Err(TruthyBoolError {
            value: other.to_string(),
        }),
    }
}

/// Error returned when [`truthy_bool`] fails to classify a value.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TruthyBoolError {
    value: String,
}

impl fmt::Display for TruthyBoolError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "unrecognised boolean value '{value}' (expected yes/y/true/1 or no/n/false/0)",
            value = self.value
        )
    }
}

impl StdError for TruthyBoolError {}

/// Trims leading and trailing whitespace before parsing a value.
///
/// `trimmed` delegates to [`FromStr`] implementations after normalising the
/// input. Errors from the inner parser are preserved.
///
/// # Examples
/// ```
/// # use rstest_bdd::datatable::trimmed;
/// let value: i32 = trimmed(" 42 ").unwrap();
/// assert_eq!(value, 42);
/// ```
///
/// # Errors
///
/// Returns [`TrimmedParseError`] when parsing the trimmed value fails.
pub fn trimmed<T>(value: &str) -> Result<T, TrimmedParseError<T::Err>>
where
    T: std::str::FromStr,
    T::Err: StdError + Send + Sync + 'static,
{
    value
        .trim()
        .parse()
        .map_err(|err| TrimmedParseError { source: err })
}

/// Error returned when [`trimmed`] fails to parse the value.
#[derive(Debug)]
pub struct TrimmedParseError<E> {
    source: E,
}

impl<E> fmt::Display for TrimmedParseError<E>
where
    E: fmt::Display,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "failed to parse trimmed value: {}", self.source)
    }
}

impl<E> StdError for TrimmedParseError<E> where E: StdError + 'static {}
