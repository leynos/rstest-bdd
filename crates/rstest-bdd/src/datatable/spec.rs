use std::collections::HashMap;
use std::error::Error as StdError;

use super::DataTableError;

/// Metadata describing the header row of a data table.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HeaderSpec {
    columns: Vec<String>,
    index: HashMap<String, usize>,
}

impl HeaderSpec {
    /// Builds a [`HeaderSpec`] from the supplied header row.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::DuplicateHeader`] when column names repeat.
    pub fn new(header_row: Vec<String>) -> Result<Self, DataTableError> {
        let mut index = HashMap::with_capacity(header_row.len());
        for (idx, column) in header_row.iter().enumerate() {
            if index.insert(column.clone(), idx).is_some() {
                return Err(DataTableError::DuplicateHeader {
                    column: column.clone(),
                });
            }
        }
        Ok(Self {
            columns: header_row,
            index,
        })
    }

    /// Returns the number of columns declared in the header.
    #[must_use]
    pub fn len(&self) -> usize {
        self.columns.len()
    }

    /// Returns `true` when the header is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.columns.is_empty()
    }

    /// Fetches a column name by index.
    #[must_use]
    pub fn column(&self, index: usize) -> Option<&str> {
        self.columns.get(index).map(String::as_str)
    }

    /// Looks up the index for a column, returning an error when missing.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingColumn`] when the column is absent.
    pub fn require(&self, name: &str, row_number: usize) -> Result<usize, DataTableError> {
        self.index
            .get(name)
            .copied()
            .ok_or_else(|| DataTableError::MissingColumn {
                row_number,
                column: name.to_string(),
            })
    }

    /// Returns the names of all columns.
    #[must_use]
    pub fn columns(&self) -> &[String] {
        &self.columns
    }
}

/// Representation of a single row within a data table.
#[derive(Debug)]
pub struct RowSpec<'h> {
    header: Option<&'h HeaderSpec>,
    row_number: usize,
    index: usize,
    cells: Vec<String>,
    indices: Vec<Option<usize>>,
}

impl<'h> RowSpec<'h> {
    pub(super) fn new(
        header: Option<&'h HeaderSpec>,
        row_number: usize,
        index: usize,
        cells: Vec<String>,
    ) -> Self {
        let indices = (0..cells.len()).map(Some).collect();
        Self {
            header,
            row_number,
            index,
            cells,
            indices,
        }
    }

    /// Returns the 1-based row number, including any header.
    #[must_use]
    pub fn row_number(&self) -> usize {
        self.row_number
    }

    /// Returns the zero-based row index relative to the data rows (excluding
    /// the header).
    #[must_use]
    pub fn index(&self) -> usize {
        self.index
    }

    /// Returns the number of cells in the row.
    #[must_use]
    pub fn len(&self) -> usize {
        self.cells.len()
    }

    /// Returns `true` when the row contains no cells.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.cells.is_empty()
    }

    /// Provides immutable access to the underlying cells.
    #[must_use]
    pub fn cells(&self) -> &[String] {
        &self.cells
    }

    /// Retrieves the cell at a given index.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingCell`] when the index is out of range.
    pub fn cell(&self, column_index: usize) -> Result<&str, DataTableError> {
        let Some(position) = self.indices.get(column_index).and_then(|p| *p) else {
            return Err(DataTableError::MissingCell {
                row_number: self.row_number,
                column_index,
            });
        };
        self.cells
            .get(position)
            .map(String::as_str)
            .ok_or(DataTableError::MissingCell {
                row_number: self.row_number,
                column_index,
            })
    }

    /// Removes and returns the cell at `column_index`.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingCell`] when the index is out of range.
    pub fn take_cell(&mut self, column_index: usize) -> Result<String, DataTableError> {
        let Some(position) = self.indices.get(column_index).and_then(|p| *p) else {
            return Err(DataTableError::MissingCell {
                row_number: self.row_number,
                column_index,
            });
        };
        let value = self.cells.remove(position);
        if let Some(slot) = self.indices.get_mut(column_index) {
            *slot = None;
        }
        for slot in self.indices.iter_mut().flatten() {
            if *slot > position {
                *slot -= 1;
            }
        }
        Ok(value)
    }

    /// Retrieves the cell associated with a header column.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingHeader`] or
    /// [`DataTableError::MissingColumn`] when the lookup fails.
    pub fn column(&self, name: &str) -> Result<&str, DataTableError> {
        let header = self.header.ok_or(DataTableError::MissingHeader)?;
        let index = header.require(name, self.row_number)?;
        self.cell(index)
    }

    /// Removes and returns the cell associated with a header column.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingHeader`] or
    /// [`DataTableError::MissingColumn`] when the lookup fails.
    pub fn take_column(&mut self, name: &str) -> Result<String, DataTableError> {
        let header = self.header.ok_or(DataTableError::MissingHeader)?;
        let index = header.require(name, self.row_number)?;
        self.take_cell(index)
    }

    /// Parses a cell using a user-supplied parser.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingCell`] when the index is out of range
    /// or [`DataTableError::CellParse`] when the parser fails.
    pub fn parse_with<F, T, E>(&self, column_index: usize, parser: F) -> Result<T, DataTableError>
    where
        F: FnOnce(&str) -> Result<T, E>,
        E: StdError + Send + Sync + 'static,
    {
        let value = self.cell(column_index)?;
        parser(value).map_err(|err| {
            let label = self
                .header
                .and_then(|h| h.column(column_index))
                .map(ToOwned::to_owned);
            DataTableError::cell_parse(self.row_number, column_index, label, err)
        })
    }

    /// Parses a named column using a user-supplied parser.
    ///
    /// # Errors
    ///
    /// Returns [`DataTableError::MissingHeader`],
    /// [`DataTableError::MissingColumn`], or
    /// [`DataTableError::CellParse`] when access or parsing fails.
    pub fn parse_column_with<F, T, E>(&self, name: &str, parser: F) -> Result<T, DataTableError>
    where
        F: FnOnce(&str) -> Result<T, E>,
        E: StdError + Send + Sync + 'static,
    {
        let header = self.header.ok_or(DataTableError::MissingHeader)?;
        let index = header.require(name, self.row_number)?;
        self.parse_with(index, parser)
    }
}
