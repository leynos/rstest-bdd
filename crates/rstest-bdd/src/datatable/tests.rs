//! Tests the datatable runtime parsing, error reporting, and helper parsers.

use std::error::Error as StdError;
use std::fmt;

use super::{DataTableError, DataTableRow, RowSpec, Rows, trimmed, truthy_bool};
use rstest::rstest;

#[derive(Debug, Clone, PartialEq, Eq)]
struct FakeError;

impl fmt::Display for FakeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("boom")
    }
}

impl StdError for FakeError {}

#[derive(Debug, PartialEq, Eq)]
struct Pair {
    first: String,
    second: i32,
}

impl DataTableRow for Pair {
    fn parse_row(mut row: RowSpec<'_>) -> Result<Self, DataTableError> {
        let first = row.take_cell(0)?;
        let second = row.parse_with(1, trimmed)?;
        Ok(Self { first, second })
    }
}

fn assert_parses_rows<T>(
    rows: Vec<Vec<String>>,
    expected_len: usize,
    expected_data: &[T],
) -> Result<(), DataTableError>
where
    T: DataTableRow + fmt::Debug + PartialEq,
{
    let parsed: Rows<T> = rows.try_into()?;
    assert_eq!(parsed.len(), expected_len);
    let data = parsed.into_iter().collect::<Vec<_>>();
    assert_eq!(data.as_slice(), expected_data);
    Ok(())
}

#[test]
fn parses_rows_without_header() {
    assert_parses_rows(
        vec![
            vec!["alice".to_string(), "1".to_string()],
            vec!["bob".to_string(), "2".to_string()],
        ],
        2,
        &[
            Pair {
                first: "alice".to_string(),
                second: 1,
            },
            Pair {
                first: "bob".to_string(),
                second: 2,
            },
        ],
    )
    .unwrap_or_else(|err| panic!("table should parse: {err}"));
}

#[derive(Debug, PartialEq, Eq)]
struct Named {
    name: String,
    active: bool,
}

impl DataTableRow for Named {
    const REQUIRES_HEADER: bool = true;

    fn parse_row(mut row: RowSpec<'_>) -> Result<Self, DataTableError> {
        let name = row.take_column("name")?;
        let active = row.parse_column_with("active", truthy_bool)?;
        Ok(Self { name, active })
    }
}

#[test]
// Intentional: shared helper keeps duplication low while scenarios diverge on
// header handling semantics.
fn parses_rows_with_header() {
    assert_parses_rows(
        vec![
            vec!["name".to_string(), "active".to_string()],
            vec!["Alice".to_string(), "yes".to_string()],
            vec!["Bob".to_string(), "no".to_string()],
        ],
        2,
        &[
            Named {
                name: "Alice".to_string(),
                active: true,
            },
            Named {
                name: "Bob".to_string(),
                active: false,
            },
        ],
    )
    .unwrap_or_else(|err| panic!("table should parse: {err}"));
}

#[test]
fn header_is_required_when_flagged() {
    let rows = vec![
        vec!["Alice".to_string(), "yes".to_string()],
        vec!["Bob".to_string(), "no".to_string()],
    ];
    let Err(err) = Rows::<Named>::try_from(rows) else {
        panic!("missing header");
    };
    assert!(matches!(
        err,
        DataTableError::MissingColumn { column, .. } if column == "name"
    ));
}

#[test]
fn missing_header_error_is_raised_for_empty_tables() {
    let rows: Vec<Vec<String>> = Vec::new();
    let Err(err) = Rows::<Named>::try_from(rows) else {
        panic!("expected missing header error");
    };
    assert!(matches!(err, DataTableError::MissingHeader));
}

#[test]
fn uneven_rows_are_rejected() {
    #[derive(Debug)]
    struct HeaderOnly;

    impl DataTableRow for HeaderOnly {
        const REQUIRES_HEADER: bool = true;

        fn parse_row(_row: RowSpec<'_>) -> Result<Self, DataTableError> {
            Ok(Self)
        }
    }

    let rows = vec![
        vec!["name".to_string()],
        vec!["alice".to_string(), "extra".to_string()],
    ];
    let Err(err) = Rows::<HeaderOnly>::try_from(rows) else {
        panic!("uneven rows");
    };
    assert!(matches!(err, DataTableError::UnevenRow { .. }));
}

#[rstest]
#[case("yes", true)]
#[case("y", true)]
#[case("true", true)]
#[case("1", true)]
#[case("no", false)]
#[case("n", false)]
#[case("false", false)]
#[case("0", false)]
fn truthy_bool_accepts_common_forms(#[case] input: &str, #[case] expected: bool) {
    match truthy_bool(input) {
        Ok(actual) => assert_eq!(actual, expected),
        Err(err) => panic!("expected recognised boolean: {err}"),
    }
}

#[test]
fn truthy_bool_rejects_unknown_values() {
    let Err(err) = truthy_bool("maybe") else {
        panic!("value is ambiguous");
    };
    assert_eq!(
        err.to_string(),
        "unrecognised boolean value 'maybe' (expected yes/y/true/1 or no/n/false/0)"
    );
}

#[rstest]
#[case(" 42 ", 42)]
#[case("\t7\n", 7)]
fn trimmed_parses_trimmed_values(#[case] input: &str, #[case] expected: i32) {
    match trimmed::<i32>(input) {
        Ok(actual) => assert_eq!(actual, expected),
        Err(err) => panic!("expected integer to parse: {err}"),
    }
}

#[test]
fn trimmed_preserves_inner_error() {
    #[derive(Debug, Clone, PartialEq, Eq)]
    struct Dummy(u8);

    impl std::str::FromStr for Dummy {
        type Err = FakeError;

        fn from_str(_s: &str) -> Result<Self, Self::Err> {
            Err(FakeError)
        }
    }

    let Err(err) = trimmed::<Dummy>("1") else {
        panic!("expected failure");
    };
    assert_eq!(
        err.to_string(),
        "failed to parse trimmed value from input '1': boom"
    );
    assert_eq!(err.original_input(), "1");
}

#[test]
fn trimmed_reports_original_input_on_parse_failure() {
    let Err(err) = trimmed::<u8>(" 300 ") else {
        panic!("expected parse failure");
    };
    assert_eq!(err.original_input(), " 300 ");
    assert!(
        err.to_string()
            .starts_with("failed to parse trimmed value from input ' 300 '")
    );
}

#[test]
fn data_table_error_messages_cover_all_variants() {
    let duplicate = DataTableError::DuplicateHeader {
        column: "name".to_string(),
    };
    assert_eq!(
        duplicate.to_string(),
        "data table header contains duplicate column 'name'"
    );

    let uneven = DataTableError::UnevenRow {
        row_number: 3,
        expected: 2,
        actual: 3,
    };
    assert_eq!(
        uneven.to_string(),
        "data table row 3 has 3 cells but expected 2"
    );

    let missing_column = DataTableError::MissingColumn {
        row_number: 5,
        column: "age".to_string(),
    };
    assert_eq!(
        missing_column.to_string(),
        "data table row 5 is missing column 'age'"
    );

    let missing_cell = DataTableError::MissingCell {
        row_number: 2,
        column_index: 4,
    };
    assert_eq!(
        missing_cell.to_string(),
        "data table row 2 is missing cell 4"
    );

    let row_parse = DataTableError::RowParse {
        row_number: 4,
        source: Box::new(FakeError),
    };
    assert_eq!(row_parse.to_string(), "row 4: boom");

    let cell_parse = DataTableError::cell_parse(3, 1, Some("age".to_string()), FakeError);
    assert_eq!(cell_parse.to_string(), "row 3, column 2 (age): boom");
}
