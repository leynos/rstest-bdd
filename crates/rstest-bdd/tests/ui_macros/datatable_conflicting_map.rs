//! Compile-fail UI test: datatable with both `map` and `try_map` must be rejected.

use rstest_bdd::datatable::{DataTableError, Rows};
use rstest_bdd_macros::{DataTable, DataTableRow};

#[derive(Debug, DataTableRow)]
struct Example {
    #[expect(dead_code, reason = "UI test keeps Example minimal")]
    name: String,
}

fn map_rows(_rows: Rows<Example>) -> Vec<String> {
    Vec::new()
}

fn try_map_rows(_rows: Rows<Example>) -> Result<Vec<String>, DataTableError> {
    Ok(Vec::new())
}

#[derive(Debug, DataTable)]
#[datatable(row = Example, map = map_rows)]
#[datatable(try_map = try_map_rows)]
struct Conflicting(Vec<String>);

fn main() {}
