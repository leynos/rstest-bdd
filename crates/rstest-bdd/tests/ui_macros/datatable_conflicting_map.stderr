error: duplicate map/try_map attribute
  --> tests/ui_macros/datatable_conflicting_map.rs:20:13
   |
20 | #[datatable(try_map = try_map_rows)]
   |             ^^^^^^^^^^^^^^^^^^^^^^
